File: src/main/java/com/indiepost/model/ImageSet.java
package com.indiepost.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.indiepost.enums.Types.ImageSize;
import com.indiepost.jackson.LocalDateTimeToUtcStringSerializer;
import org.hibernate.annotations.*;
import org.hibernate.annotations.CascadeType;

import javax.persistence.*;
import javax.persistence.Entity;
import javax.persistence.Table;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.Set;

/**
 * Created by jake on 9/7/16.
 */
@Entity
@Table(name = "ImageSets")
public class ImageSet implements Serializable {

    private static final long serialVersionUID = -6176638635434014551L;

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Fetch(FetchMode.JOIN)
    @OneToMany(orphanRemoval = true, fetch = FetchType.EAGER)
    @Cascade({CascadeType.ALL, CascadeType.SAVE_UPDATE})
    @JoinColumn(name = "imageSetId")
    @BatchSize(size = 20)
    @JsonIgnore
    private Set<Image> images;

    @Column(nullable = false)
    @Size(min = 9, max = 10)
    private String contentType;

    @Size(max = 300)
    private String caption;

    @Size(max = 20)
    @Column(unique = true, nullable = false)
    private String prefix;

    @JsonSerialize(using = LocalDateTimeToUtcStringSerializer.class)
    @Column(nullable = false)
    private LocalDateTime uploadedAt;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Set<Image> getImages() {
        return images;
    }

    public void setImages(Set<Image> images) {
        this.images = images;
    }

    public Image getOriginal() {
        return findByImageSize(ImageSize.ORIGINAL);
    }

    public Image getLarge() {
        return findByImageSize(ImageSize.LARGE);
    }

    public Image getOptimized() {
        return findByImageSize(ImageSize.OPTIMIZED);
    }

    public Image getSmall() {
        return findByImageSize(ImageSize.SMALL);
    }

    public Image getThumbnail() {
        return findByImageSize(ImageSize.THUMBNAIL);
    }

    public String getCaption() {
        return caption;
    }

    public void setCaption(String caption) {
        this.caption = caption;
    }

    public LocalDateTime getUploadedAt() {
        return uploadedAt;
    }

    public void setUploadedAt(LocalDateTime uploadedAt) {
        this.uploadedAt = uploadedAt;
    }

    public String getContentType() {
        return contentType;
    }

    public void setContentType(String contentType) {
        this.contentType = contentType;
    }

    private Image findByImageSize(ImageSize sizeType) {
        if (images == null) {
            return null;
        }
        for (Image image : images) {
            if (image.getSizeType() == sizeType) {
                return image;
            }
        }
        return null;
    }

    public String getPrefix() {
        return prefix;
    }

    public void setPrefix(String prefix) {
        this.prefix = prefix;
    }
}

File: src/main/kotlin/com/indiepost/model/ImageSet.kt
package com.indiepost.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.indiepost.enums.Types.ImageSize
import org.hibernate.annotations.*
import org.hibernate.annotations.CascadeType
import java.io.Serializable
import java.time.LocalDateTime
import javax.persistence.*
import javax.persistence.Entity
import javax.persistence.Table
import javax.validation.constraints.Size

/**
 * Created by jake on 9/7/16.
 */
@Entity
@Table(name = "ImageSets")
data class ImageSet(
        @Id
        @GeneratedValue(strategy = GenerationType.AUTO)
        var id: Long? = null,

        @Fetch(FetchMode.JOIN)
        @OneToMany(orphanRemoval = true, fetch = FetchType.EAGER)
        @Cascade(CascadeType.ALL, CascadeType.SAVE_UPDATE)
        @JoinColumn(name = "imageSetId")
        @BatchSize(size = 20)
        @JsonIgnore
        var images: Set<Image>? = null,

        @Column(nullable = false)
        @Size(min = 9, max = 10)
        var contentType: String? = null,

        @Size(max = 300)
        var caption: String? = null,

        @Size(max = 20)
        @Column(unique = true, nullable = false)
        var prefix: String? = null,

        @Column(nullable = false)
        var uploadedAt: LocalDateTime? = null
) : Serializable {
    val original: Image?
        get() = findByImageSize(ImageSize.ORIGINAL)

    val large: Image?
        get() = findByImageSize(ImageSize.LARGE)

    val optimized: Image?
        get() = findByImageSize(ImageSize.OPTIMIZED)

    val small: Image?
        get() = findByImageSize(ImageSize.SMALL)

    val thumbnail: Image?
        get() = findByImageSize(ImageSize.THUMBNAIL)

    private fun findByImageSize(sizeType: ImageSize): Image? {
        if (images == null) {
            return null
        }
        for (image in images!!) {
            if (image.sizeType == sizeType) {
                return image
            }
        }
        return null
    }

    companion object {
        private const val serialVersionUID = -6176638635434014551L
    }
}
